### PR TITLE
(go look at https://github.com/lhr-solar/Sunlight/pull/34)    Initialization (and integration)

### PR DESCRIPTION
**Pull request by tiandahuang**
_Note: integration of all tasks and integration with data acquisition is currently untested, but is ready to merge once tested_

Adds initialization to the tasks and integrates all the tasks. Tasks are synchronized after initialization with a counting semaphore. 
Also includes a few fixes and CAN changes and a bit of cleanup.

Next step (optional) is to add basic error handling.

### CODE REVIEW

File: Core/Inc/config.h [hunk 1]
```diff
@@ -31,11 +38,17 @@
 
 // Helper Macros
 
+#ifdef LWIP_HDR_DEF_H // def.h
```
Comment (manthanand) [inline Core/Inc/config.h:41]: should the define for this also be located in this file?

File: Core/Src/DataLoggingTask.c [hunk 2]
```diff
@@ -1,17 +1,33 @@
...
     int cntr = 0;
+    if (SDCard_Init() != FR_OK);
+    xSemaphoreTake(InitSem, 0);
+    while(uxSemaphoreGetCount(InitSem) != 0);
 
     while (1){
         if (SDCard_Sort_Write_Data() == FR_OK) cntr++; //increment counter if data was written
         if (cntr > MOUNTCYCLES) {
-            if (SDCard_CloseFileSystem() != FR_OK); //TODO: ERROR CHECKING HERE
+            if (SDCard_CloseFileSystem() != FR_OK);
```
Comment (manthanand) [inline Core/Src/DataLoggingTask.c]: I would get rid of the if statement and add a comment saying if the error occurs, there isn't much we can do about it.

**Comment by manthanand** [comment]
It might be a good idea if our error handling for things we can't deal with only ran in debugging mode. So we could still have the if statements but they would only run if we were debugging (and we could print stuff for things that didn't work). Maybe for another PR though.

File: Core/Src/DataLoggingTask.c (hunk 2 above)
Comment (tiandahuang) [reply Core/Src/DataLoggingTask.c]: I could add a function that spins somewhere, like the BPS fault state. Maybe I could make use of the LEDs on the board?

File: Core/Inc/config.h (hunk 1 above)
Comment (tiandahuang) [reply Core/Inc/config.h:41]: This is just for the ethernet stuff; the ```#ifdef``` makes sure that the macro and static function only exist in files that need it.

File: Core/Src/DataLoggingTask.c (hunk 2 above)
Comment (manthanand) [reply Core/Src/DataLoggingTask.c]: That would be nice. We can save it for another PR.

File: Core/Inc/main.h [hunk 3]
```diff
@@ -49,8 +49,14 @@ extern SPI_HandleTypeDef hspi5;
 extern UART_HandleTypeDef huart1;
 extern UART_HandleTypeDef huart3;
 
-extern ETH_HandleTypeDef heth; //
-extern TIM_HandleTypeDef htim1;//
+extern ETH_HandleTypeDef heth;
+extern TIM_HandleTypeDef htim1;
+
+// for debugging: dropped messages count
+extern uint32_t GPSDroppedMessages;
```
Comment (manthanand) [inline Core/Inc/main.h]: Why are these here? Are they used outside of their libraries?

File: Drivers/Periphs/Inc/radio.h [hunk 4]
```diff
@@ -1,16 +1,20 @@
 #ifndef RADIO_H
 #define RADIO_H
 
+/**
+ * @file radio.h
```
Comment (manthanand) [inline Drivers/Periphs/Inc/radio.h:2]: All the copyright headers should go above the ifndef

File: Core/Inc/main.h (hunk 3 above)
Comment (tiandahuang) [reply Core/Inc/main.h]: Thought maybe we could use them somewhere to track how much data we're losing; I should probably stick that in as #ifdef debug but let me know if it sounds useful

Comment (manthanand) [reply Core/Inc/main.h]: I think if it is a debug utility, we would add the debugging functionality in their respective library. Tbh, I just don't want too many shared things, even if they are atomic.

File: Drivers/Periphs/Src/IMU.c [hunk 5]
```diff
@@ -225,15 +225,20 @@ HAL_StatusTypeDef IMU_Init(){
...
- * @return HAL_StatusTypeDef - OK, ERROR, BUSY, or TIMEOUT
+ * @return HAL_OK on success, IMU_BUSY if called too soon after last call, HAL_StatusTypeDef != HAL_OK on error
  */
 HAL_StatusTypeDef IMU_GetMeasurements(IMUData_t *Data){
     // Read 18 contiguous data registers within the IMU starting at 0x08 and stores data within struct fields
-    //wait 10ms before collecting data again
-    static int ms = portMAX_DELAY;
-    while (ms + 10 > xTaskGetTickCount());
+    
+    static TickType_t ticks = 0;
```
Comment (manthanand) [inline Drivers/Periphs/Src/IMU.c:233]: You

File: Core/Src/BroadcastingTask.c [hunk 6]
```diff
@@ -13,13 +13,16 @@
...
+#include "config.h"
 #include <stdio.h>
 
 void BroadcastingTask(void *argument){
     
     Ethernet_Init();
+    debugprintf("Ethernet: connected to server\n\r");
 
     while (1){
+        printf("c");
```
Comment (manthanand) [inline Core/Src/BroadcastingTask.c:25]: What is this for?

Comment (manthanand) [reply Core/Src/BroadcastingTask.c:25]: Should it be a debugprintf?

Comment (tiandahuang) [reply Core/Src/BroadcastingTask.c:25]: don't worry about it i'm gonna take it out later 👮
